ntoskrnl/ps/process.c

```diff
@@ -1614,4 +1677,234 @@ NtOpenProcess(OUT PHANDLE ProcessHandle,
+{
+    return !is_backwards
+               ? PsGetNextProcess(current)
+               : PsGetPreviousProcess(current);
+}
+
+NTSTATUS
+NTAPI
+NtGetNextProcess(
+    _In_ HANDLE ProcessHandle,
+    _In_ ACCESS_MASK DesiredAccess,
+    _In_ ULONG HandleAttributes,
+    _In_ ULONG Flags,
+    _Out_ PHANDLE NewProcessHandle
+)
+{
+    const KPROCESSOR_MODE previousMode = KeGetPreviousMode();
+
+    /* Validate user attributes */
+    HandleAttributes = ObpValidateAttributes(HandleAttributes, previousMode);
+
+    /* Check if we came from user mode */
+    if (previousMode != KernelMode)
+    {
+        _SEH2_TRY
+        {
+            /* Probe process handle */
+            ProbeForWriteHandle(NewProcessHandle);
+
+            *NewProcessHandle = NULL;
+        }
+        _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+        {
+            /* Return the exception code */
+            _SEH2_YIELD(return _SEH2_GetExceptionCode());
+        }
+        _SEH2_END;
+    }
+    else
+    {
+        *NewProcessHandle = NULL;
+    }
+
+    const BOOLEAN is_backwards = Flags & 1;
+    Flags ^= Flags & 1;
+    if (Flags)
+    {
+        DPRINT1("NtGetNextProcess flags (%#X) unsupported");
+        return STATUS_INVALID_PARAMETER;
+    }
+
+    PEPROCESS current_process = NULL;
+    if (ProcessHandle)
+    {
+        /* Reference it */
+        const NTSTATUS status = ObReferenceObjectByHandle(ProcessHandle,
+                                                          0,
+                                                          PsProcessType,
+                                                          previousMode,
+                                                          (PVOID*)&current_process,
+                                                          NULL);
+        if (!NT_SUCCESS(status))
+        {
+            DPRINT1("NtGetNextProcess ObReferenceObjectByHandle -> %#X", status);
+            _SEH2_YIELD(return status);
+        }
+    }
+
+    for (PEPROCESS target = NtpGetAdjacentProcess(is_backwards, current_process);
+         target;
+         target = NtpGetAdjacentProcess(is_backwards, target))
+    {
+        current_process = NULL;
+        _SEH2_TRY
+        {
+            ACCESS_STATE accessState;
+            AUX_ACCESS_DATA auxData;
+            {
+                /* Create an access state */
+                const NTSTATUS status = SeCreateAccessState(&accessState,
+                                                            &auxData,
+                                                            DesiredAccess,
+                                                            &PsProcessType->TypeInfo.GenericMapping);
+                if (!NT_SUCCESS(status))
+                    return status;
+
+                /* Check if this is a debugger */
+                if (SeSinglePrivilegeCheck(SeDebugPrivilege, previousMode))
+                {
+                    /* Did he want full access? */
```

Comment:
The caller is not a he, it's either an "it" or a "they".

Reply:
Copied from some function above.